flink-table/flink-table-planner/src/main/java/org/apache/calcite/rex/RexSimplify.java

```diff
@@ -2024,7 +2029,9 @@ private RexNode simplifySearch(RexCall call, RexUnknownAs unknownAs) {
     private RexNode simplifyCast(RexCall e) {
         RexNode operand = e.getOperands().get(0);
         operand = simplify(operand, UNKNOWN);
-        if (sameTypeOrNarrowsNullability(e.getType(), operand.getType())) {
+        if (sameTypeOrNarrowsNullability(e.getType(), operand.getType())
```

Comment:
The change in this line and in `RexLiteral` is a blocker for this PR in my opinion. We should fix these things in Calcite first and only port the changes to the Flink code base, once merged in Calcite.

Reply:
I've spent quite some time here, and couldn't find a workaround.

I could modify the PR and only allow it for TableAPI for the moment, and try to fix this in Calcite first.
If it's accepted, have a follow up PR to enable this for SQL in flink.